llsd/base.py

```diff
@@ -483,19 +477,197 @@ def _parse_string_delim(self, delim):
             insert_idx += 1
 
         # Sync our local read index with the canonical one
-        self._index = read_idx
         try:
             # Slice off only what we used of the working decode buffer
             return decode_buff[:insert_idx].decode('utf-8')
         except UnicodeDecodeError as exc:
             self._error(exc)
 
 
+class BytesUtil:
+    """
+    Utility methods implemented on an input bytes object
+    """
+    __slots__ = ['_buffer', '_index']
+
+    def __init__(self, input):
+        self._buffer = input
+        self._index = 0
+
+    def error(self, message, offset=0):
+        try:
+            byte = chr(self._buffer[self._index+offset])
+        except IndexError:
+            byte = None
+        raise LLSDParseError("%s at byte %d: %r" % (message, self._index+offset, byte))
+
+    def peek_impl(self, num, full=True):
+        # full=True means error if we can't peek ahead num bytes
+        if full and self._index + num > len(self._buffer):
+            self.error("Trying to read past end of buffer")
+        return self._buffer[self._index:self._index + num]
+
+    def getc(self, num=1, full=True):
+        # full=True means error if we can't read num bytes
+        chars = self.peek_impl(num, full)
+        self._index += num
+        return chars
+
+
+class StreamUtil:
```

Comment:
Instead of building wrappers around bytes and streams, could `llsd.parse` convert all input into a unified class such as `BufferedReader` which provides the requisite peek implementation? This would make these classes unneeded.

Reply:
+1 interested to hear discussion about this as well

Reply:
The distinction between `StreamUtil` and `BytesUtil` becomes moot when `LLSDBaseParser.force_stream` is `True`, which it is by default: in that case, instead of using `BytesUtil` at all, `_reset()` copies the input `bytes` object into an `io.BytesIO` and proceeds with `StreamUtil`.

I was curious whether a `BytesUtil` wrapper would have any performance benefits over the tactic above. I suspect it only matters if the consumer passes a very large `bytes` object, in which case both the buffer allocation and the copy itself would incur noticeable overhead. But - in the likely case that the consumer is reading that very large `bytes` object from a stream, we could address that overhead by passing the stream directly to `parse()`.

Bottom line: we can probably make `BytesUtil` go away entirely. In that case, yes, we could wrap the input stream (whether passed by the caller or an internal `BytesIO`) in a `BufferedReader` and use its `peek()`. It might still be useful to wrap it in a class that provides `StreamUtil.error()`.

Is that the direction we should go?

Reply:
`BytesStream` and `AnyStream` could similarly be elided.

Reply:
I'd like to see SteamUtil, MatchStream, AnyStream, BytesStream disappeared, they're confusing and, imo, a too complicated when we can just take input and coerce it into a standard I/O contract.